skimage/restoration/_cycle_spin.py

```diff
@@ -0,0 +1,160 @@
+    func : function
+        A function to apply to circularly shifted versions of ``x``.  Should
+        take ``x`` as its first argument. Any additional arguments can be
+        supplied via ``func_kw``.
+    max_shifts : int or tuple
+        If an integer, shifts in ``range(0, max_shifts+1)`` will be used along
+        each axis of ``x``. If a tuple, ``range(0, max_shifts[i]+1)`` will be
+        along axis i.
+    shift_steps : int or tuple, optional
+        The step size for the shifts applied along axis, i, are::
+        ``range((0, max_shifts[i]+1, shift_steps[i]))``. If an integer is
+        provided, the same step size is used for all axes.
+    num_workers : int or None, optional
+        The number of parallel threads to use during cycle spinning. If set to
+        ``None``, the full set of available cores are used.
+    multichannel : bool, optional
+        Whether to treat the final axis as channels (no cycle shifts are
+        performed over the channels axis).
+    func_kw : dict, optional
+        Additional keyword arguments to supply to ``func``.
+
+    Returns
+    -------
+    avg_y : np.ndarray
+        The output of ``func(x, **func_kw)`` averaged over all combinations of
+        the specified axis shifts.
+
+    Notes
+    -----
+    Cycle spinning was proposed as a way to approach shift-invariance via
+    performing several circular shifts of a shift-variant transform [1]_.
+
+    For a n-level discrete wavelet transforms, one may wish to perform all
+    shifts up to ``max_shifts = 2**n - 1``. In practice, much of the benefit
+    can often be realized with only a small number of shifts per axis.
+
+    For transforms such as the blockwise discrete cosine transform, one may
+    wish to evaluate shifts up to the block size used by the transform.
+
+    References
+    ----------
+    .. [1] R.R. Coifman and D.L. Donoho.  "Translation-Invariant De-Noising".
+           Wavelets and Statistics, Lecture Notes in Statistics, vol.103.
+           Springer, New York, 1995, pp.125-150.
+           DOI:10.1007/978-1-4612-2544-7_9
+
+    Examples
+    --------
+    >>> import skimage.data
+    >>> from skimage import img_as_float
+    >>> from skimage.restoration import denoise_wavelet, cycle_spin
+    >>> img = img_as_float(skimage.data.camera())
+    >>> sigma = 0.1
+    >>> img = img + sigma * np.random.standard_normal(img.shape)
+    >>> denoised = cycle_spin(img, func=denoise_wavelet, max_shifts=3)
+
+    """
+    x = np.asanyarray(x)
+    all_shifts = _generate_shifts(x.ndim, multichannel, max_shifts,
+                                  shift_steps)
+    all_shifts = list(all_shifts)
+
+    def _run_one_shift(shift):
+        # shift, apply function, inverse shift
+        xs = _roll_axes(x, shift)
+        tmp = func(xs, **func_kw)
+        return _roll_axes(tmp, -np.asarray(shift))
+
+    # compute a running average across the cycle shifts
+    avg_y = 0
+    if num_workers == 1:
```

Comment:
What's the overhead for simply using dask in this case as well?

Reply:
The example (which currently uses num_workers=1 implicitly) takes ~4.7 s to run on this system with the codes as it is now.  This increases to 5.2 s to run if I modify to just use dask in all cases.  so ~ 10% overhead in that case.  

Also, now that dask is a requirement we could default `num_workers=None` in `cycle_spin` to have multithreading by default and the user can set num_workers=1 if desired.  What do you think?